File: modules/expo-widgets/android/src/main/java/expo/modules/widgets/ExpoWidgetsModule.kt
package expo.modules.widgets

import android.appwidget.AppWidgetManager
import android.appwidget.AppWidgetProvider
import android.content.ComponentName
import android.content.Context
import android.content.Intent
import expo.modules.kotlin.exception.Exceptions
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.kotlin.records.Field
import expo.modules.kotlin.records.Record
import org.json.JSONArray
import org.json.JSONObject

class PendingSmsRecord : Record {
  @Field val id: String = ""
  @Field val merchant: String = ""
  @Field val amount: String = ""
  @Field val currency: String = ""
  @Field val sender: String = ""
  @Field val date: Long? = null
  @Field val processingStatus: String = "pending"
}

class ExpoWidgetsModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("ExpoWidgets")

    AsyncFunction("syncWidgetData") { snapshot: Map<String, Any?> ->
      val context: Context = appContext.reactContext ?: throw Exceptions.ReactContextLost()
      syncSnapshot(context, snapshot)
      refreshWidgetProviders(context)
    }

    AsyncFunction("refreshWidgets") {
      val context: Context = appContext.reactContext ?: throw Exceptions.ReactContextLost()
      refreshWidgetProviders(context)
    }

    AsyncFunction("storePendingSms") { records: List<PendingSmsRecord> ->
      val context: Context = appContext.reactContext ?: throw Exceptions.ReactContextLost()
      storePendingSmsRecords(context, records)
      refreshWidgetProviders(context)
    }
  }

  private fun storePendingSmsRecords(context: Context, records: List<PendingSmsRecord>) {
    val prefs = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)

    // Filter only pending records and sort by recency (newest date first; null date -> 0L)
    val pendingRecords = records
      .filter { it.processingStatus == "pending" }
      .sortedByDescending { it.date ?: 0L }

    // Collect existing record IDs (for cleanup of stale entries)
    val existingJson = prefs.getString(KEY_PENDING_SMS_RECORDS, null)
    val previousIds = mutableSetOf<String>()
    if (existingJson != null) {
      try {
        val arr = JSONArray(existingJson)
        for (i in 0 until arr.length()) {
          previousIds.add(arr.getString(i))
        }
      } catch (e: Exception) {
        android.util.Log.e("ExpoWidgets", "Failed to parse existing pending SMS records", e)
      }
    }

    val currentIds = mutableListOf<String>()
    val cleanIds = mutableSetOf<String>()

    prefs.edit().apply {
      for (record in pendingRecords) {
        val id = record.id
        if (id.isEmpty()) continue
        currentIds.add(id)
        cleanIds.add(id)

        putString("${SmsQuickImportReceiver.EXTRA_PREFIX}${id}_merchant", record.merchant)
        putString("${SmsQuickImportReceiver.EXTRA_PREFIX}${id}_amount", record.amount)
        putString("${SmsQuickImportReceiver.EXTRA_PREFIX}${id}_currency", record.currency)
        putString("${SmsQuickImportReceiver.EXTRA_PREFIX}${id}_sender", record.sender)
        putLong("${SmsQuickImportReceiver.EXTRA_PREFIX}${id}_date", record.date ?: 0L)
      }

      // Store the ordered list of record IDs as JSON array
      putString(KEY_PENDING_SMS_RECORDS, JSONArray(currentIds).toString())

      // Clean up stale entries that are no longer in the new records
      val staleIds = previousIds - cleanIds
      for (staleId in staleIds) {
        remove("${SmsQuickImportReceiver.EXTRA_PREFIX}${staleId}_merchant")
        remove("${SmsQuickImportReceiver.EXTRA_PREFIX}${staleId}_amount")
        remove("${SmsQuickImportReceiver.EXTRA_PREFIX}${staleId}_currency")
        remove("${SmsQuickImportReceiver.EXTRA_PREFIX}${staleId}_sender")
        remove("${SmsQuickImportReceiver.EXTRA_PREFIX}${staleId}_date")
      }
    }.apply()
  }

  private fun syncSnapshot(context: Context, snapshot: Map<String, Any?>) {
    val prefs = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
    val safeToSpend = snapshot["safeToSpend"] as? Map<*, *>
    val theme = snapshot["theme"] as? Map<*, *>
    val streak = snapshot["streak"] as? Map<*, *>
    @Suppress("UNCHECKED_CAST")
    val pendingSms = snapshot["pendingSms"] as? List<Map<*, *>>
    val pet = snapshot["pet"] as? Map<*, *>

    prefs.edit().apply {
      // ---- safeToSpend ------------------------------------------------
      if (safeToSpend == null) {
        remove(KEY_SAFE_TO_SPEND_AMOUNT)
        remove(KEY_SAFE_TO_SPEND_CURRENCY)
        remove(KEY_SAFE_TO_SPEND_FORMATTED_AMOUNT)
        remove(KEY_SAFE_TO_SPEND_TITLE)
        remove(KEY_SAFE_TO_SPEND_SUBTITLE)
        remove(KEY_SAFE_TO_SPEND_UPDATED_AT)
      } else {
        putFloat(KEY_SAFE_TO_SPEND_AMOUNT, (safeToSpend["amount"] as? Number)?.toFloat() ?: 0f)
        putString(KEY_SAFE_TO_SPEND_CURRENCY, safeToSpend["currencyCode"] as? String ?: "")
        putString(KEY_SAFE_TO_SPEND_FORMATTED_AMOUNT, safeToSpend["formattedAmount"] as? String ?: "")
        putString(KEY_SAFE_TO_SPEND_TITLE, safeToSpend["title"] as? String ?: "")
        putString(KEY_SAFE_TO_SPEND_SUBTITLE, safeToSpend["subtitle"] as? String ?: "")
        putLong(KEY_SAFE_TO_SPEND_UPDATED_AT, (safeToSpend["updatedAt"] as? Number)?.toLong() ?: System.currentTimeMillis())
      }

      // ---- theme -------------------------------------------------------
      if (theme == null) {
        remove(KEY_THEME_ID)
        remove(KEY_THEME_MODE)
        remove(KEY_THEME_BACKGROUND_START)
        remove(KEY_THEME_BACKGROUND_END)
        remove(KEY_THEME_TITLE_COLOR)
        remove(KEY_THEME_PRIMARY_TEXT_COLOR)
        remove(KEY_THEME_SECONDARY_TEXT_COLOR)
        remove(KEY_THEME_ACTION_ICON_COLOR)
        remove(KEY_THEME_INCOME_ACCENT_COLOR)
        remove(KEY_THEME_EXPENSE_ACCENT_COLOR)
        remove(KEY_THEME_TRANSFER_ACCENT_COLOR)
      } else {
        putString(KEY_THEME_ID, theme["themeId"] as? String ?: "")
        putString(KEY_THEME_MODE, theme["themeMode"] as? String ?: "")
        putString(KEY_THEME_BACKGROUND_START, theme["backgroundStartColor"] as? String ?: "")
        putString(KEY_THEME_BACKGROUND_END, theme["backgroundEndColor"] as? String ?: "")
        putString(KEY_THEME_TITLE_COLOR, theme["titleColor"] as? String ?: "")
        putString(KEY_THEME_PRIMARY_TEXT_COLOR, theme["primaryTextColor"] as? String ?: "")
        putString(KEY_THEME_SECONDARY_TEXT_COLOR, theme["secondaryTextColor"] as? String ?: "")
        putString(KEY_THEME_ACTION_ICON_COLOR, theme["actionIconColor"] as? String ?: "")
        putString(KEY_THEME_INCOME_ACCENT_COLOR, theme["incomeAccentColor"] as? String ?: "")
        putString(KEY_THEME_EXPENSE_ACCENT_COLOR, theme["expenseAccentColor"] as? String ?: "")
        putString(KEY_THEME_TRANSFER_ACCENT_COLOR, theme["transferAccentColor"] as? String ?: "")
      }

      putBoolean(KEY_IS_PRIVACY_ENABLED, snapshot["isPrivacyEnabled"] as? Boolean ?: false)

      // ---- streak ------------------------------------------------------
      if (streak == null) {
        remove(KEY_STREAK_COUNT)
        remove(KEY_STREAK_TODAY_LOGGED)
        remove(KEY_STREAK_LAST_LOGGED_DATE)
        remove(KEY_STREAK_CAN_RECOVER)
        remove(KEY_STREAK_MISSED_DAYS)
      } else {
        putInt(KEY_STREAK_COUNT, (streak["count"] as? Number)?.toInt() ?: 0)
        putBoolean(KEY_STREAK_TODAY_LOGGED, streak["todayLogged"] as? Boolean ?: false)
        putString(KEY_STREAK_LAST_LOGGED_DATE, streak["lastLoggedDate"] as? String ?: "")
        putBoolean(KEY_STREAK_CAN_RECOVER, streak["canRecover"] as? Boolean ?: false)
        putInt(KEY_STREAK_MISSED_DAYS, (streak["missedDays"] as? Number)?.toInt() ?: 0)
      }

      // ---- pendingSms --------------------------------------------------
      if (pendingSms == null) {
        remove(KEY_PENDING_SMS_COUNT)
        for (i in 0 until MAX_PENDING_SMS) {
          remove("${KEY_PENDING_SMS_PREFIX}${i}_id")
          remove("${KEY_PENDING_SMS_PREFIX}${i}_merchant")
          remove("${KEY_PENDING_SMS_PREFIX}${i}_amount")
          remove("${KEY_PENDING_SMS_PREFIX}${i}_currency")
        }
      } else {
        val count = minOf(pendingSms.size, MAX_PENDING_SMS)
        putInt(KEY_PENDING_SMS_COUNT, count)
        for (i in 0 until count) {
          val item = pendingSms[i]
          val amountNum = (item["amount"] as? Number)?.toDouble() ?: 0.0
          val amountStr = String.format(java.util.Locale.US, "%.2f", amountNum)
          putString("${KEY_PENDING_SMS_PREFIX}${i}_id", item["id"] as? String ?: "")
          putString("${KEY_PENDING_SMS_PREFIX}${i}_merchant", item["merchant"] as? String ?: "")
          putString("${KEY_PENDING_SMS_PREFIX}${i}_amount", amountStr)
          putString("${KEY_PENDING_SMS_PREFIX}${i}_currency", item["currency"] as? String ?: "")
        }
        // Clear stale slots
        for (i in count until MAX_PENDING_SMS) {
          remove("${KEY_PENDING_SMS_PREFIX}${i}_id")
          remove("${KEY_PENDING_SMS_PREFIX}${i}_merchant")
          remove("${KEY_PENDING_SMS_PREFIX}${i}_amount")
          remove("${KEY_PENDING_SMS_PREFIX}${i}_currency")
        }
      }

      // ---- pet ---------------------------------------------------------
      if (pet == null) {
        remove(KEY_PET_HEALTH)
        remove(KEY_PET_MOOD)
        remove(KEY_PET_LEVEL)
      } else {
        putInt(KEY_PET_HEALTH, (pet["health"] as? Number)?.toInt() ?: 50)
        putString(KEY_PET_MOOD, pet["mood"] as? String ?: "happy")
        putInt(KEY_PET_LEVEL, (pet["level"] as? Number)?.toInt() ?: 1)
      }
    }.apply()
  }

  private fun refreshWidgetProviders(context: Context) {
    val appWidgetManager = AppWidgetManager.getInstance(context)

    WIDGET_PROVIDER_CLASS_NAMES.forEach { className ->
      runCatching {
        val providerClass = Class
          .forName("${context.packageName}.$className")
          .asSubclass(AppWidgetProvider::class.java)

        val componentName = ComponentName(context, providerClass)
        val ids = appWidgetManager.getAppWidgetIds(componentName)
        if (ids.isEmpty()) {
          return@runCatching
        }

        val updateIntent = Intent(context, providerClass).apply {
          action = AppWidgetManager.ACTION_APPWIDGET_UPDATE
          putExtra(AppWidgetManager.EXTRA_APPWIDGET_IDS, ids)
        }
        context.sendBroadcast(updateIntent)
      }
    }
  }

  companion object {
    const val PREFS_NAME = "full_frills_balance_widgets"
    private const val MAX_PENDING_SMS = 5

    // safeToSpend
    private const val KEY_SAFE_TO_SPEND_AMOUNT = "safe_to_spend_amount"
    private const val KEY_SAFE_TO_SPEND_CURRENCY = "safe_to_spend_currency"
    private const val KEY_SAFE_TO_SPEND_FORMATTED_AMOUNT = "safe_to_spend_formatted_amount"
    private const val KEY_SAFE_TO_SPEND_TITLE = "safe_to_spend_title"
    private const val KEY_SAFE_TO_SPEND_SUBTITLE = "safe_to_spend_subtitle"
    private const val KEY_SAFE_TO_SPEND_UPDATED_AT = "safe_to_spend_updated_at"

    // theme
    private const val KEY_THEME_ID = "widget_theme_id"
    private const val KEY_THEME_MODE = "widget_theme_mode"
    private const val KEY_THEME_BACKGROUND_START = "widget_theme_background_start"
    private const val KEY_THEME_BACKGROUND_END = "widget_theme_background_end"
    private const val KEY_THEME_TITLE_COLOR = "widget_theme_title_color"
    private const val KEY_THEME_PRIMARY_TEXT_COLOR = "widget_theme_primary_text_color"
    private const val KEY_THEME_SECONDARY_TEXT_COLOR = "widget_theme_secondary_text_color"
    private const val KEY_THEME_ACTION_ICON_COLOR = "widget_theme_action_icon_color"
    private const val KEY_THEME_INCOME_ACCENT_COLOR = "widget_theme_income_accent_color"
    private const val KEY_THEME_EXPENSE_ACCENT_COLOR = "widget_theme_expense_accent_color"
    private const val KEY_THEME_TRANSFER_ACCENT_COLOR = "widget_theme_transfer_accent_color"

    // privacy
    private const val KEY_IS_PRIVACY_ENABLED = "widget_is_privacy_enabled"
    private const val KEY_PENDING_SMS_RECORDS = "pending_sms_records"

    // streak
    private const val KEY_STREAK_COUNT = "streak_count"
    private const val KEY_STREAK_TODAY_LOGGED = "streak_today_logged"
    private const val KEY_STREAK_LAST_LOGGED_DATE = "streak_last_logged_date"
    private const val KEY_STREAK_CAN_RECOVER = "streak_can_recover"
    private const val KEY_STREAK_MISSED_DAYS = "streak_missed_days"

    // pendingSms
    private const val KEY_PENDING_SMS_COUNT = "pending_sms_count"
    private const val KEY_PENDING_SMS_PREFIX = "pending_sms_"

    // pet
    private const val KEY_PET_HEALTH = "pet_health"
    private const val KEY_PET_MOOD = "pet_mood"
    private const val KEY_PET_LEVEL = "pet_level"

    val WIDGET_PROVIDER_CLASS_NAMES = listOf(
      "JournalLauncherWidgetProvider",
      "SafeToSpendWidgetProvider",
      "SafeToSpendActionsWidgetProvider",
      "SafeToSpendActionsSquareWidgetProvider",
      "StreakPetWidgetProvider",
      "SmsTriageWidgetProvider",
    )
  }
}

File: modules/expo-widgets/ios/ExpoWidgetsModule.swift
import ExpoModulesCore
import WidgetKit
import OSLog

private let logger = Logger(subsystem: "in.sahilsoni.fullfrillsbalance", category: "ExpoWidgets")

struct PendingSmsRecord: Record {
  @Field var id: String = ""
  @Field var merchant: String = ""
  @Field var amount: String = ""
  @Field var currency: String = ""
  @Field var sender: String = ""
  @Field var date: Double?
  @Field var processingStatus: String = "pending"
}

public final class ExpoWidgetsModule: Module {

  public func definition() -> ModuleDefinition {
    Name("ExpoWidgets")

    AsyncFunction("syncWidgetData") { (snapshot: [String: Any]) in
      guard let defaults = UserDefaults(suiteName: WidgetConstants.appGroupId) else {
        logger.error("Failed to open UserDefaults for app group")
        return
      }

      // ---- safeToSpend ------------------------------------------------
      if let safeToSpend = snapshot["safeToSpend"] as? [String: Any] {
        defaults.set((safeToSpend["amount"] as? NSNumber)?.doubleValue ?? 0, forKey: "safe_to_spend_amount")
        defaults.set(safeToSpend["currencyCode"] as? String ?? "", forKey: "safe_to_spend_currency")
        defaults.set(safeToSpend["formattedAmount"] as? String ?? "", forKey: "safe_to_spend_formatted_amount")
        defaults.set(safeToSpend["title"] as? String ?? "", forKey: "safe_to_spend_title")
        defaults.set(safeToSpend["subtitle"] as? String ?? "", forKey: "safe_to_spend_subtitle")
        defaults.set((safeToSpend["updatedAt"] as? NSNumber)?.doubleValue ?? Date().timeIntervalSince1970 * 1000, forKey: "safe_to_spend_updated_at")
      } else {
        defaults.removeObject(forKey: "safe_to_spend_amount")
        defaults.removeObject(forKey: "safe_to_spend_currency")
        defaults.removeObject(forKey: "safe_to_spend_formatted_amount")
        defaults.removeObject(forKey: "safe_to_spend_title")
        defaults.removeObject(forKey: "safe_to_spend_subtitle")
        defaults.removeObject(forKey: "safe_to_spend_updated_at")
      }

      // ---- theme -------------------------------------------------------
      if let theme = snapshot["theme"] as? [String: Any] {
        defaults.set(theme["themeId"] as? String ?? "", forKey: "widget_theme_id")
        defaults.set(theme["themeMode"] as? String ?? "", forKey: "widget_theme_mode")
        defaults.set(theme["backgroundStartColor"] as? String ?? "", forKey: "widget_theme_background_start")
        defaults.set(theme["backgroundEndColor"] as? String ?? "", forKey: "widget_theme_background_end")
        defaults.set(theme["titleColor"] as? String ?? "", forKey: "widget_theme_title_color")
        defaults.set(theme["primaryTextColor"] as? String ?? "", forKey: "widget_theme_primary_text_color")
        defaults.set(theme["secondaryTextColor"] as? String ?? "", forKey: "widget_theme_secondary_text_color")
        defaults.set(theme["actionIconColor"] as? String ?? "", forKey: "widget_theme_action_icon_color")
        defaults.set(theme["incomeAccentColor"] as? String ?? "", forKey: "widget_theme_income_accent_color")
        defaults.set(theme["expenseAccentColor"] as? String ?? "", forKey: "widget_theme_expense_accent_color")
        defaults.set(theme["transferAccentColor"] as? String ?? "", forKey: "widget_theme_transfer_accent_color")
      } else {
        defaults.removeObject(forKey: "widget_theme_id")
        defaults.removeObject(forKey: "widget_theme_mode")
        defaults.removeObject(forKey: "widget_theme_background_start")
        defaults.removeObject(forKey: "widget_theme_background_end")
        defaults.removeObject(forKey: "widget_theme_title_color")
        defaults.removeObject(forKey: "widget_theme_primary_text_color")
        defaults.removeObject(forKey: "widget_theme_secondary_text_color")
        defaults.removeObject(forKey: "widget_theme_action_icon_color")
        defaults.removeObject(forKey: "widget_theme_income_accent_color")
        defaults.removeObject(forKey: "widget_theme_expense_accent_color")
        defaults.removeObject(forKey: "widget_theme_transfer_accent_color")
      }

      // ---- privacy -----------------------------------------------------
      if let isPrivacyEnabled = snapshot["isPrivacyEnabled"] as? Bool {
        defaults.set(isPrivacyEnabled, forKey: "widget_is_privacy_enabled")
      } else {
        defaults.set(false, forKey: "widget_is_privacy_enabled")
      }

      // ---- streak ------------------------------------------------------
      if let streak = snapshot["streak"] as? [String: Any] {
        defaults.set(streak["count"] as? Int ?? 0, forKey: "streak_count")
        defaults.set(streak["todayLogged"] as? Bool ?? false, forKey: "streak_today_logged")
        defaults.set(streak["lastLoggedDate"] as? String ?? "", forKey: "streak_last_logged_date")
        defaults.set(streak["canRecover"] as? Bool ?? false, forKey: "streak_can_recover")
        defaults.set(streak["missedDays"] as? Int ?? 0, forKey: "streak_missed_days")
      } else {
        defaults.removeObject(forKey: "streak_count")
        defaults.removeObject(forKey: "streak_today_logged")
        defaults.removeObject(forKey: "streak_last_logged_date")
        defaults.removeObject(forKey: "streak_can_recover")
        defaults.removeObject(forKey: "streak_missed_days")
      }

      // ---- pet ---------------------------------------------------------
      if let pet = snapshot["pet"] as? [String: Any] {
        defaults.set(pet["health"] as? Int ?? 100, forKey: "pet_health")
        defaults.set(pet["mood"] as? String ?? "happy", forKey: "pet_mood")
      } else {
        defaults.removeObject(forKey: "pet_health")
        defaults.removeObject(forKey: "pet_mood")
      }

      // ---- pendingSms --------------------------------------------------
      if let pendingSms = snapshot["pendingSms"] as? [[String: Any]] {
        let count = min(pendingSms.count, 5)
        defaults.set(count, forKey: "pending_sms_count")
        var currentIds: [String] = []

        for i in 0..<count {
          let item = pendingSms[i]
          let id = item["id"] as? String ?? "\(i)"
          let merchant = item["merchant"] as? String ?? ""
          let amount = (item["amount"] as? NSNumber)?.doubleValue ?? 0
          let amountStr = String(format: "%.2f", amount)
          let currency = item["currency"] as? String ?? ""

          currentIds.append(id)

          // Set indexed keys
          defaults.set(id, forKey: "pending_sms_\(i)_id")
          defaults.set(merchant, forKey: "pending_sms_\(i)_merchant")
          defaults.set(amountStr, forKey: "pending_sms_\(i)_amount")
          defaults.set(currency, forKey: "pending_sms_\(i)_currency")

          // Set PendingSmsKeys contract
          defaults.set(merchant, forKey: PendingSmsKeys.merchant(for: id))
          defaults.set(amountStr, forKey: PendingSmsKeys.amount(for: id))
          defaults.set(currency, forKey: PendingSmsKeys.currency(for: id))
        }

        if let encoded = try? JSONEncoder().encode(currentIds) {
          defaults.set(encoded, forKey: PendingSmsKeys.records)
        }

        // Clear stale slots beyond the current count
        for i in count..<5 {
          defaults.removeObject(forKey: "pending_sms_\(i)_id")
          defaults.removeObject(forKey: "pending_sms_\(i)_merchant")
          defaults.removeObject(forKey: "pending_sms_\(i)_amount")
          defaults.removeObject(forKey: "pending_sms_\(i)_currency")
        }
      } else {
        defaults.removeObject(forKey: "pending_sms_count")
        defaults.removeObject(forKey: PendingSmsKeys.records)
        for i in 0..<5 {
          defaults.removeObject(forKey: "pending_sms_\(i)_id")
          defaults.removeObject(forKey: "pending_sms_\(i)_merchant")
          defaults.removeObject(forKey: "pending_sms_\(i)_amount")
          defaults.removeObject(forKey: "pending_sms_\(i)_currency")
        }
      }

      // ---- pet ---------------------------------------------------------
      if let pet = snapshot["pet"] as? [String: Any] {
        defaults.set(pet["health"] as? Int ?? 50, forKey: "pet_health")
        defaults.set(pet["mood"] as? String ?? "happy", forKey: "pet_mood")
        defaults.set(pet["level"] as? Int ?? 1, forKey: "pet_level")
      } else {
        defaults.removeObject(forKey: "pet_health")
        defaults.removeObject(forKey: "pet_mood")
        defaults.removeObject(forKey: "pet_level")
      }

      if #available(iOS 14.0, *) {
        WidgetCenter.shared.reloadAllTimelines()
      }
    }

    AsyncFunction("refreshWidgets") {
      if #available(iOS 14.0, *) {
        WidgetCenter.shared.reloadAllTimelines()
      }
    }

    /// Store pending SMS records for the SMS Triage widget.
    /// Called from JS after the SMS pipeline produces pending items that need
    /// user approval. Serialises each record into AppGroup UserDefaults using
    /// the key pattern `pending_sms_<id>_<field>` so the widget extension can
    /// read them without accessing the database or waking JS.
    AsyncFunction("storePendingSms") { (records: [PendingSmsRecord]) in
      guard let defaults = UserDefaults(suiteName: WidgetConstants.appGroupId) else {
        return
      }

      // Filter only pending records and sort by recency (newest date first; null date -> 0.0)
      let pendingRecords = records
        .filter { $0.processingStatus == "pending" }
        .sorted { ($0.date ?? 0.0) > ($1.date ?? 0.0) }

      // Collect all existing record IDs (for cleanup of stale entries)
      let existingRecordsData = defaults.data(forKey: PendingSmsKeys.records)
      let decoder = JSONDecoder()
      var previousIds = Set<String>()
      if let existingData = existingRecordsData {
        do {
          let ids = try decoder.decode([String].self, from: existingData)
          previousIds = Set(ids)
        } catch {
          os_log(.error, "[ExpoWidgets] Failed to decode pending_sms_records: %{public}@", error.localizedDescription)
        }
      }

      var currentIds: [String] = []
      var cleanIds = Set<String>()

      for record in pendingRecords {
        let id = record.id
        guard !id.isEmpty else { continue }
        currentIds.append(id)
        cleanIds.insert(id)

        defaults.set(record.merchant, forKey: PendingSmsKeys.merchant(for: id))
        defaults.set(record.amount, forKey: PendingSmsKeys.amount(for: id))
        defaults.set(record.currency, forKey: PendingSmsKeys.currency(for: id))
        defaults.set(record.sender, forKey: PendingSmsKeys.sender(for: id))
        defaults.set(record.date ?? 0.0, forKey: PendingSmsKeys.date(for: id))
      }

      // Store the ordered list of record IDs
      let encoder = JSONEncoder()
      do {
        let encoded = try encoder.encode(currentIds)
        defaults.set(encoded, forKey: PendingSmsKeys.records)
      } catch {
        os_log(.error, "[ExpoWidgets] Failed to encode pending_sms_records: %{public}@", error.localizedDescription)
      }

      // Clean up stale entries that are no longer in the new records
      let staleIds = previousIds.subtracting(cleanIds)
      for staleId in staleIds {
        defaults.removeObject(forKey: PendingSmsKeys.merchant(for: staleId))
        defaults.removeObject(forKey: PendingSmsKeys.amount(for: staleId))
        defaults.removeObject(forKey: PendingSmsKeys.currency(for: staleId))
        defaults.removeObject(forKey: PendingSmsKeys.sender(for: staleId))
        defaults.removeObject(forKey: PendingSmsKeys.date(for: staleId))
      }

      if #available(iOS 14.0, *) {
        WidgetCenter.shared.reloadAllTimelines()
      }
    }
  }
}

